__tests__/integration/testUtils.js

```diff
@@ -0,0 +1,123 @@
+const knex = require("../../server/database/knex");
+const uuid = require('uuid');
+const log = require("loglevel");
+const Crypto = require('crypto');
+const generator = require('generate-password');
+const JWTService = require("../../server/services/JWTService");
+const Transfer = require("../../server/models/Transfer");
+const { expect } = require('chai');
+
+/*
+ * register the user, create password hash, and apiKey
+ */
+async function register(user){
+  const sha512 = function(password, salt){
+    var hash = Crypto.createHmac('sha512', salt); /** Hashing algorithm sha512 */
+    hash.update(password);
+    var value = hash.digest('hex');
+    return value;
+  };
+
+  const salt = Crypto.randomBytes(32).toString('base64')  //create a secure salt
+  const passwordHash = sha512(user.password, salt)
+
+  const apiKey = generator.generate({
+      length: 32,
+      numbers: true
+  });
+
+  await knex('api_key')
+    .insert({
+      key: apiKey,
+      tree_token_api_access: true,
+      hash: 'test',
+      salt: 'test',
+      name: 'test',
+    });
+
+  // wallet
+  const result = await knex('wallet')
+    .insert({
+      id: uuid.v4(),
+      name: user.name,
+      password: passwordHash,
+      salt: salt,
+    }).returning("*");
+  log.info("registered wallet:", result);
+  return {
+    ...result[0], 
+    apiKey, 
+    //restore password
+    password: user.password,
+  };
+}
+
+/*
+ * create the user, apiKey, then login, return
+ * token
+ */
+async function registerAndLogin(user){
+  const userRegistered = await register(user);
+  const jwtService = new JWTService();
+  const token = jwtService.sign(userRegistered);
+  userRegistered.token = token;
+  expect(userRegistered).property("apiKey").a("string");
+  return userRegistered;
+}
+
+async function clear() {
+  log.debug('clear tables');
+  await knex('api_key').del();
+  await knex('transaction').del();
+  await knex('token').del();
+  await knex('wallet').del();
+  await knex('wallet_trust').del();
+  await knex('transfer').del();
+}
+
+/*
+ * Add a token to a wallet
+ */
+async function addToken(wallet, token){
```

Comment:
Compare to use a real endpoint to prepare the data, use Knex is more straightforward and stable, I think because we are coding, keep changing the endpoint API, so it's fragile to use them here.

Reply:
So, are you suggesting to use direct access to database in order to check data correctness during end-to-end test or do you think it's better to use api and treat the repo as a "black box"?

Reply:
My thoughts on this is that:
For integration tests, use functions that directly visit the DB, in this way, it's simple and straightforward.
For End to End test, we should use a real API/endpoint to treat the whole app as a block box.
Do you agree with this?

Reply:
yeah, I think this makes sense